odoo/http.py

```diff
@@ -415,6 +368,133 @@ def serialize_exception(exception):
         'context': getattr(exception, 'context', {}),
     }
 
+
+# =========================================================
+# File Streaming
+# =========================================================
+
+KEEP = object()  # sentinel
+
+@dataclass
+class StreamData:
+    type = ''  # 'data' or 'path' or 'url'
+    data = None
+    path = None
+    url = None
+
+    mimetype = None
+    as_attachment = False
+    download_name = None
+    conditional = True
+    etag = True
+    last_modified = None
+    max_age = None
+    size = None
+
+    @classmethod
+    def from_path(cls, path):
+        path = file_path(path)
+        return cls(
+            type='path',
+            path=path,
+            as_attachment=as_attachment,
+            size=os.stat(path).size
+        )
+
+    @classmethod
+    def from_attachment(cls, attachment):
+        self = cls(
+            mimetype=attachment.mimetype,
+            as_attachment=True,
+            download_name=attachment.name,
+            conditional=True,
+            etag=attachment.checksum,
+            max_age=STATIC_CACHE_LONG,
+        )
+
+        if attachment.stored_fname:
+            stat = os.stat
+            self.type = 'path'
+            self.path = werkzeug.security.safe_join(
+                os.path.abspath(config.filestore()),
+                attachment.stored_fname
+            )
+            self.last_modified = stat.mtime
+            self.size = stat.size
+
+        elif attachment.db_datas:
+            self.type = 'data'
+            self.data = attachment.raw
+            self.size = len(self.data)
+
+        elif attachment.url:
+            self.type = 'url'
+            self.url = attachment.url
+
+            # When the URL targets a file located in an addon, assume it
+            # is a path instead. It saves a 3xx response.
+            if (attachment.url or '').startswith('/'):
+                with contextlib.suppress(FileNotFoundError):
+                    return cls.from_path(attachment.url)
+
+        return self
+
+    @classmethod
+    def from_image_attachment(cls, attachment, resized_image_data=None):
+        self = cls.from_attachment(attachment)
+        self.as_attachment = False
+        self.etag = f'{attachment.checksum}-{self.size}'
+
+        if resized_image_data:
+            self.type = 'data'
+            self.path = None
+            self.data = resized_image_data
+            self.etag = f'{attachment.checksum}-{len(resized_image_data)}'
+
+    @wraps(werkzeug_send_file)
+    def make_http_response(self, **send_file_kwargs):
+        assert self.type in ('url', 'data', 'path') and getattr(self, self.type) is not None
```

Comment:
Should have useful error messages. There should probably be errors in the factory methods when no `type` could be decided.